--- 
title: Project 1
subtitle: Week 3 (on example counts)
output: html_document
---

**QC Metrics for Full Data**  
STAR alignment shows that around 90% of each sample is uniquely mapped which indicates good alignment. FastQC results show 
high sequence quality, normal distribution of GC content, and very low percentage of over-represented sequences (1% of reads). However, 
there does seem to be an unusual level of sequence duplication. Sequence duplication means that some reads are repeated many times. This could indicate problems during library preparation (over-amplification) or these genes are possibly highly expressed. Overall, the sequencence quality is good.

```{r include = FALSE}
library(tidyverse)
library(DESeq2)
library(RColorBrewer)
library(pheatmap)
```

**Filtering**  
The counts contains 2000 genes and 9 samples. Genes that had a total expression less than 5 were filtered out which resulted in 875 genes. 
```{r}
counts <- read_csv("counts_example.csv")

counts_filtered <- counts %>%
  filter(rowSums(select(., where(is.numeric))) >= 5)
```

```{r echo=FALSE}
# table of filter summary
num_samples <- ncol(counts) - 1
total_num_genes <- nrow(counts)
filter_num_genes <- nrow(counts_filtered)
percent_pass <- (filter_num_genes / total_num_genes) * 100
not_pass_genes <- total_num_genes - filter_num_genes
percent_not_pass <- (not_pass_genes / total_num_genes) * 100

sum_counts_table <- tibble(
  Measure = c("Number of Samples", 
              "Total Number of Genes",
              "Number of genes passing filter",
              "% of genes passing filter",
              "Number of genes not passing",
              "% of genes not passing"
  ),
  Value = c(num_samples,
            total_num_genes,
            filter_num_genes,
            paste0(percent_pass, "%"),
            not_pass_genes,
            paste0(percent_not_pass, "%"))
)
print(sum_counts_table)
```


**Differential Analysis**  

```{r include=FALSE}
# counts matrix without gene ids
count_mat <- as.matrix(counts_filtered[-1])

# gene ids
row.names(count_mat) <- counts_filtered$gene

# create a sample matrix from sample names
sample_info <- tibble(
  sample_name=colnames(counts_filtered[-1])
) %>%
  separate(sample_name,c("timepoint","replicate"),sep="_",remove=FALSE) %>%
  mutate(
    timepoint=factor(timepoint,levels=c("AD","P0","P4","P7"))
  )

design <- formula(~ timepoint)

dds <- DESeqDataSetFromMatrix(
  countData=count_mat,
  colData=sample_info,
  design=design
)
dds <- DESeq(dds)
resultsNames(dds)

res <- results(dds, name="timepoint_P0_vs_AD")
```

Top 10 significant genes ranked by padj

```{r}
p0_vs_Ad_de <- as_tibble(res) %>%
  mutate(gene=rownames(res)) %>%
  relocate(gene) %>%
  arrange(pvalue)
head(p0_vs_Ad_de, 10)
```

*Gene names from the parsed GTF file will be added later.*  

At a padj threshold of 0.5, 214 genes are significant.  
```{r}
sign <- filter(p0_vs_Ad_de,padj<0.05)
```

```{r include=FALSE}
# removing decimals from list of genes to upload into DAVID
gene_list <- list(sign$gene)
gene_list_no_decimal <- lapply(gene_list, function(gene_vector) {
  sub("\\.\\d+$", "", gene_vector)
})
writeLines(unlist(gene_list_no_decimal), "genes.txt")
```

34 clusters were identified by DAVID. The functional annotation clustering results showed high enrichment (6.38 enrichment score) in genes related to mitochondrial function in comparison to other functional clusters such as relating to flavin adenine dinucleotide binding or ATPase (both around 1.8 enrichment score).  

**RNAseq Quality Control Plots**  

```{r}
# normalization
rld <- rlog(dds)

# PCA
plotPCA(rld, intgroup=c("timepoint", "replicate"))

```

```{r echo=FALSE}
# heatmap of the sample-to-sample distances
sampleDists <- dist(t(assay(rld)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(rld$timepoint, rld$replicate, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "RdBu")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
```

From the PCA plot, we can see 4 distinct clusters based on the timepoints. The two AD samples, the two P0 samples, the two P4 samples, and the two P7 samples form 4 clusters. We can see that PC1 explains about 67% of the variance. The heatmap shows that each of the replicates at each timepoint are very similar. Then, the P7 samples and the P4 samples are close. Then, the P7 and P4 samples are close to the P0 samples. Finally, the P7, P4, and P0 samples are close to the AD samples. This indicates that the AD samples are the least similar to the other samples. It also seems like P0 and P7 have a smaller distance than P0 and P4 judging by the light orange color. 
